### ISSUE TITLE
What do you think of changing the name of the plugin to sf-metadata-transformer-plugin or just metadata-transformer-plugin

### ISSUE THREAD

**Issue by AllanOricil** [mcarvin8/sfdx-decomposer-plugin#17]
Inside this plugin there would exist default metadata transformers:

- xml-composer
- xml-decomposer
...
- any other metadata transformation it could exist in the future

Then we can think of a way to create transformation pipelines for any type of Metadata.

````
metadata in -> transformation 1 -> transformation 2 -> ... -> metadata out
````

This way if someone decides to extend it and create its own transformation for a specific metadata type, they can just add a new transformer in the pipeline

// post-retrieve pipeline
````
metadata in -> xml decomposer -> metadata out
````

// pre-deploy pipeline
````
metadata in -> xml composer -> metadata out
````

This project would ship standard xml composer and xml decomposer packages, as well as a pipeline framework.

**Comment by mcarvin8** [comment]
Interesting to see another plugin that does it for profiles.

I think decomposing/recomposing, splitting, or transforming can convey the same idea for the most part.

That splitter profile plugin can transform XML to JSON data, which is an interesting concept. 

If we wanted to expand this to transform the XML into different formats like JSON, then transformer probably makes the most sense for the name.

But I'm not sure the benefits of transforming it to JSON when the metadata API takes XMLs. Guess the benefit is JSON being easier to read than an XML in a Git repo. Unless there is some future initiative by Salesforce to allow the API to accept JSONs.

**Comment by AllanOricil** [comment]
oh I was not talking about transforming file types. I used the word "transfom" to mean any type of transformation but keeping the file type that the metadata api works with. That sfdx-profile-splitter could later be added in the xml-composer | xml-decomposer transformers. Another reason for using the word "transform" is because Node uses it for [pipeline transformers](https://nodejs.org/api/stream.html#streampipelinesource-transforms-destination-options), which are made to transform files while buffering them. 

We could of course support the convertion of xml -> json building another transformer that would be shipped with this plugin.

// xml to json pipeline
````bash
metadata input -> xml / json transformer -> metadata output
````

**Comment by mcarvin8** [comment]
I'm fine with renaming this project to `sf-metadata-transformer` before releasing v1.0.0.

I would like to get @scolladon 's PR merged into `beta` for the async updates once it's ready.

Afterward, I can update the `package.json` name on beta to `sf-metadata-transformer`, release a new Beta version, and confirm the 2nd beta is stable.

Then, I will merge `beta` into `main` to build v1.0.0. Once `main` is at v1.0.0 with the new name, I'll rename this repo.

**Comment by AllanOricil** [comment]
@mcarvin8 I just thought again and I believe you should keep the name as is for v1. If you change it now, it would become a compromise to make this complex project, and I don't know how long or how much effort it would take to put it together. Later, if more people are interested we can think about it renaming it.

**Comment by mcarvin8** [comment]
Sure @AllanOricil . I'm fine with leaving as-is just to keep the scope simple for now. We can see interest later on for expanding it and then revisit this. 

I am planning on starting the standalone XML decomposer plugin package soon since that definitely could be added soon for anyone sick of reviewing large XMLs. 

The fast xml parsing classes to decompose and compose XMLs on this beta release are pretty isolated right now. The only SF specific thing I see right now in them is the "field names" used to name decompose files, which could be changed to something generic like a hash like you suggested. 

It's the file handler classes which are really Salesforce metadata specific.

**Comment by AllanOricil** [comment]
@mcarvin8 you could try to use my [js-template](https://github.com/AllanOricil/js-template) to write this lib and save some time with configs (don't forget to give me a star)

**Comment by scolladon** [comment]
My 2 cents (again) I really like @AllanOricil suggestions/ideas (as most of the times ;) ) and I think @mcarvin8 you should keep an organized backlog to track those. I personally use github [project](https://docs.github.com/en/issues/planning-and-tracking-with-projects/learning-about-projects/about-projects) to track cards and prioritize everything.

From what I see, I think the plugin is pretty stable and valuable now. Technically I would say that short term priority should be around maintainability (hexagonal architecture, unit test, integration test, functional test, e2e test, CI/CD automation). Having a high level of maintainability will ease the addition of new feature suggested by @AllanOricil without breaking existing customers.

**Comment by mcarvin8** [comment]
Thank you @AllanOricil ! I was just about to ask if you had any good templates to reference for package building!

**Comment by mcarvin8** [comment]
Great points @scolladon . I'll create something to track suggestions like this and tech backlog items for future releases.

Agree that I want to keep the plugin stable and maintainable as we scale it. Definitely some more work to do in improving maintainability like adding more tests into the pipeline that can easily check new features against a solid baseline.

**Comment by mcarvin8** [comment]
Closing this discussion but will add to a issue tracker.

**Comment by mcarvin8** [comment]
Funny enough - Salesforce's SDR repo has transformer and decomposer in 1 file-path lol
![image](https://github.com/mcarvin8/sfdx-decomposer-plugin/assets/90224411/c56b1aba-b89d-4d08-9025-84f2dda694c8)

**Comment by AllanOricil** [comment]
Whoever wrote that probably knew about node stream transforms too. I think their plan was to do what we discussed above, and maybe they gave up, or don't have enought resources. You could ask someone from the cli team if they would be willing to help in this project.

**Comment by mcarvin8** [comment]
The general XML disassembler package has been shipped with this plug-in's v3.0.0 after spending the last week or so testing and confirming the updates needed to back-port this change. 

This takes out a big chunk of this plug-in's code base so that this plugin focuses on handling the right Salesforce files and the general XML disassembler handles the heavy lifting. 

Thank you Allan for the suggestion! Good call!

**Comment by mcarvin8** [comment]
This has been renamed to `sf-decomposer`.

This package does include additional XML disassembler child packages which do xml2json and xml2yaml conversions for those who want YAML/JSON decomposed files.
